flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -241,28 +241,131 @@ Fortran::lower::FirOpBuilder::createShape(mlir::Location loc,
       [&](const fir::ArrayBoxValue &box) { return consShape(loc, box); },
       [&](const fir::CharArrayBoxValue &box) { return consShape(loc, box); },
       [&](const fir::BoxValue &box) { return consShape(loc, box); },
+      [&](const fir::IrBoxValue &box) -> mlir::Value {
+        if (!box.getLBounds().empty()) {
+          auto shiftType =
+              fir::ShiftType::get(getContext(), box.getLBounds().size());
+          return create<fir::ShiftOp>(loc, shiftType, box.getLBounds());
+        }
+        return mlir::Value{};
+      },
       [&](auto) -> mlir::Value { fir::emitFatalError(loc, "not an array"); });
 }
 
+mlir::Value
+Fortran::lower::FirOpBuilder::readCharLen(mlir::Location loc,
+                                          const fir::ExtendedValue &box) {
+  return box.match(
+      [&](const fir::CharBoxValue &x) -> mlir::Value { return x.getLen(); },
+      [&](const fir::CharArrayBoxValue &x) -> mlir::Value {
+        return x.getLen();
+      },
+      [&](const fir::IrBoxValue &x) -> mlir::Value {
+        assert(x.isCharacter());
+        if (!x.getExplicitParameters().empty())
+          return x.getExplicitParameters()[0];
+        return Fortran::lower::CharacterExprHelper{*this, loc}
+            .readLengthFromBox(x.getAddr());
```

Comment:
Aside: We've pruned some of these builder helpers down a lot. I wonder if we can eliminate any of them at this point.

Reply:
You mean `CharacterExprHelper` ? It is still doing quite a lot around characters, the best we could do I think is to refactor the class out of the interface. It's not really useful to have it as a class outside of the implementation.
e.g. here it would be simpler to have: `Fortran::lower::readCharLengthFromBox(*this, loc, x.getAddr());`